pkg/kv/kvserver/replica_evaluate_test.go

```diff
@@ -244,6 +249,97 @@ func TestEvaluateBatch(t *testing.T) {
 				require.Equal(t, "value-e", string(b))
 			},
 		},
+		//
+		// Test suite for KeyLocking.
+		//
+		{
+			// Three scans that observe 3, 1, and 0 keys, respectively. An
+			// unreplicated lock should be acquired on each key that is scanned.
+			name: "scan with key locking",
+			setup: func(t *testing.T, d *data) {
+				writeABCDEF(t, d)
+				d.ba.Txn = &txn
+				scanAD := scanArgsString("a", "d")
+				scanAD.KeyLocking = lock.Exclusive
+				d.ba.Add(scanAD)
+				scanEF := scanArgsString("e", "f")
+				scanEF.KeyLocking = lock.Exclusive
+				d.ba.Add(scanEF)
+				scanHJ := scanArgsString("h", "j")
+				scanHJ.KeyLocking = lock.Exclusive
+				d.ba.Add(scanHJ)
+			},
+			check: func(t *testing.T, r resp) {
+				verifyScanResult(t, r, []string{"a", "b", "c"}, []string{"e"}, nil)
+				verifyAcquiredLocks(t, r, lock.Unreplicated, "a", "b", "c", "e")
+				verifyAcquiredLocks(t, r, lock.Replicated, []string(nil)...)
+			},
+		},
+		{
+			// Ditto in reverse.
```

Comment:
Test the interaction of scans that a) have a limit (=acquire only locks for keys returned) and b) return tombstones (do we put locks on the tombstone keys and is this intentional? Add commentary on the flag that turns tombstone retrieval on).